include/veriblock/storage/block_provider.hpp

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2019-2020 Xenios SEZC
+// https://www.veriblock.org
+// Distributed under the MIT software license, see the accompanying
+// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef VERIBLOCK_POP_CPP_BLOCK_PROVIDER_HPP
+#define VERIBLOCK_POP_CPP_BLOCK_PROVIDER_HPP
+
+#include <memory>
+
+#include "veriblock/blockchain/block_index.hpp"
+#include "veriblock/entities/altblock.hpp"
+#include "veriblock/entities/btcblock.hpp"
+#include "veriblock/entities/vbkblock.hpp"
+#include "veriblock/storage/block_iterator.hpp"
+#include "veriblock/validation_state.hpp"
+
+namespace altintegration {
+
+/**
+ * @struct BlockProvider
+ *
+ * An abstraction over on-disk storage iterator.
+ *
+ * veriblock-pop-cpp does not dictate how to iterate over the storage. Altchain
+ * must create derived class.
+ *
+ * @ingroup interfaces
+ */
+
+template <typename BlockT>
+struct BlockProvider {
```

Comment:
I think this interface should be hidden in namespace `details` and then there would be another interface which implements this interface specific for each block, e.g.
```
struct BlockProvider : public GenericBlockProvider<AltBlock>, public GenericBlockProvider<VbkBlock>, public GenericBlockProvider<BtcBlock> {}
```